### PR TITLE
 feat(metrics): add experiment data to amplitude user properties 

### LINKED ISSUE #5542: Implement experiment and experiment_variation properties on amplitude events
`experiment` and ~~`experiment_version`~~ `experiment_variation` should be added to `user_properties` when they're available.

### PR DESCRIPTION
**Pull request by philbooth**
Fixes #5542.

`WIP` until I've chatted to @davismtl about the implementation, no need to review yet.

```
{"op":"amplitudeEvent","time":1507558148822,"device_id":"68399ce840544a4880450e51a05158d8","event_type":"fxa_reg - view","session_id":1507558147898,"event_properties":{"device_id":"68399ce840544a4880450e51a05158d8","entrypoint":"menupanel","service":"sync"},"user_properties":{"flow_id":"a01b77ecc31ec7bc39f4e78fa735ec1ba9f6533f9e2e9394990d24e8953571ac","experiment_q3form_changes":"signup_password_confirm","experiment_signup_password_confirm":"control"},"app_version":"97","language":"en"}
{"op":"amplitudeEvent","time":1507558149875,"device_id":"68399ce840544a4880450e51a05158d8","event_type":"fxa_reg - engage","session_id":1507558147898,"event_properties":{"device_id":"68399ce840544a4880450e51a05158d8","entrypoint":"menupanel","service":"sync"},"user_properties":{"flow_id":"a01b77ecc31ec7bc39f4e78fa735ec1ba9f6533f9e2e9394990d24e8953571ac","experiment_q3form_changes":"signup_password_confirm","experiment_signup_password_confirm":"control"},"app_version":"97","language":"en"}
{"op":"amplitudeEvent","time":1507558155639,"device_id":"68399ce840544a4880450e51a05158d8","event_type":"fxa_reg - submit","session_id":1507558147898,"event_properties":{"device_id":"68399ce840544a4880450e51a05158d8","entrypoint":"menupanel","service":"sync"},"user_properties":{"flow_id":"a01b77ecc31ec7bc39f4e78fa735ec1ba9f6533f9e2e9394990d24e8953571ac","experiment_q3form_changes":"signup_password_confirm","experiment_signup_password_confirm":"control"},"app_version":"97","language":"en"}
{"op":"amplitudeEvent","time":1507558232942,"user_id":"b6fc9f2a18d84249b149730295de5bd9","device_id":"124282c74bf24f789353736f246bacc7","event_type":"fxa_email - click","session_id":1507558147898,"event_properties":{"device_id":"124282c74bf24f789353736f246bacc7","email_type":"registration","service":"sync"},"user_properties":{"flow_id":"a01b77ecc31ec7bc39f4e78fa735ec1ba9f6533f9e2e9394990d24e8953571ac","experiment_q3form_changes":"signup_password_confirm","experiment_signup_password_confirm":"control","experiment_send_sms":"treatment"},"app_version":"97","language":"en"}
{"op":"amplitudeEvent","time":1507558233039,"user_id":"b6fc9f2a18d84249b149730295de5bd9","device_id":"124282c74bf24f789353736f246bacc7","event_type":"fxa_sms - view","session_id":1507558147898,"event_properties":{"device_id":"124282c74bf24f789353736f246bacc7"},"user_properties":{"flow_id":"a01b77ecc31ec7bc39f4e78fa735ec1ba9f6533f9e2e9394990d24e8953571ac","experiment_q3form_changes":"signup_password_confirm","experiment_signup_password_confirm":"control","experiment_send_sms":"treatment"},"app_version":"97","language":"en"}
```

### CODE REVIEW

**Comment by davismtl** [comment]
Hey @philbooth ,

I'm a bit confused by:
- experiment_q3form_changes
- experiment_signup_password_confirm

It seems like this could get messy and pollute our list of user properties.

Seems like we should just have the following user properties:
- experiment_name (value equal to: signup_password_confirm)
- experiment_variation (value equal to: control)

But even my suggestion isn't perfect because it tends to only remember the last value of the properties so if a user is in 2 tests, we risk losing the original values.

I'm asking Amplitude in Slack what their recommendation is for this.

**Comment by rfk** [comment]
Do we have any guarantee that a user will always be in the same bucket for the same experiment.  For example, if we have an experiment that changes the sign-in flow, I can imagine a user getting one variation when signing in on device A, and a different variation when signing in on device B.  We can't pin the experiment to the user because we don't know who they are until after they sign in.

**Comment by philbooth** [comment]
> Do we have any guarantee that a user will always be in the same bucket for the same experiment. For example, if we have an experiment that changes the sign-in flow, I can imagine a user getting one variation when signing in on device A, and a different variation when signing in on device B. We can't pin the experiment to the user because we don't know who they are until after they sign in.

Yep, this is an issue. One solution would be to emit experiment events like we do for the flow metrics, which would have the advantage of tying each experiment to a device id, session id and timestamp. A less ideal but still workable solution would be to use the `$append` syntax to enable multiple states per experiment, which would mean losing that extra metadata.

**Comment by davismtl** [comment]
I asked Amplitude:

> What’s best practice for passing A/B test variations into amplitude?
>I was thinking of passing test name and variation as a user property but user properties usually reflect the last value only. How would I make sure that we keep traces of all the tests a user has been a part of?

From Amplitude:

> User properties are applied to subsequent events but filtering on those properties in the UI filters on the property value at the time the event was performed.
> 
> Our recommendation is to pass along a/b tests as a user property rather than discrete events.

This seems better than I thought.

>Do we have any guarantee that a user will always be in the same bucket for the same experiment.

With Amplitude's recommendations, a user could be part of each variation so I think we shouldn't run into this much.

>Yep, this is an issue. One solution would be to emit experiment events like we do for the flow metrics, which would have the advantage of tying each experiment to a device id, session id and timestamp. A less ideal but still workable solution would be to use the $append syntax to enable multiple states per experiment, which would mean losing that extra metadata.

Based on what Amplitude answered, I don't think any of this is needed.

**Comment by philbooth** [comment]
> Based on what Amplitude answered, I don't think any of this is needed.

Sounds good, thanks @davismtl.

Do we still need to have a conversation about the taxonomy or is the implementation in this PR along the right lines now?

**Comment by philbooth** [comment]
> Do we still need to have a conversation about the taxonomy...

For the record the outcome of that conversation is captured in https://github.com/mozilla/fxa-content-server/issues/5542#issuecomment-335532554, I'm updating this PR to match what's said there.

**Comment by philbooth** [comment]
Updated as per https://github.com/mozilla/fxa-content-server/issues/5542#issuecomment-335532554:

```
{"op":"amplitudeEvent","time":1507736081879,"device_id":"60ad83df377244a18bbc67b8a2307cc3","event_type":"fxa_reg - view","session_id":1507736080954,"event_properties":{"device_id":"60ad83df377244a18bbc67b8a2307cc3","entrypoint":"menupanel","service":"sync"},"user_properties":{"flow_id":"1a0b1cc2636b4f8198b05cbfb126a1aadda754c8c1e132490524f94897d3c7d8","experiments":["disabled_button_state_treatment","q3form_changes_disabled_button_state"]},"app_version":"97","language":"en"}
{"op":"amplitudeEvent","time":1507736083127,"device_id":"60ad83df377244a18bbc67b8a2307cc3","event_type":"fxa_reg - engage","session_id":1507736080954,"event_properties":{"device_id":"60ad83df377244a18bbc67b8a2307cc3","entrypoint":"menupanel","service":"sync"},"user_properties":{"flow_id":"1a0b1cc2636b4f8198b05cbfb126a1aadda754c8c1e132490524f94897d3c7d8","experiments":["disabled_button_state_treatment","q3form_changes_disabled_button_state"]},"app_version":"97","language":"en"}
{"op":"amplitudeEvent","time":1507736089016,"device_id":"60ad83df377244a18bbc67b8a2307cc3","event_type":"fxa_reg - submit","session_id":1507736080954,"event_properties":{"device_id":"60ad83df377244a18bbc67b8a2307cc3","entrypoint":"menupanel","service":"sync"},"user_properties":{"flow_id":"1a0b1cc2636b4f8198b05cbfb126a1aadda754c8c1e132490524f94897d3c7d8","experiments":["disabled_button_state_treatment","q3form_changes_disabled_button_state"]},"app_version":"97","language":"en"}
{"op":"amplitudeEvent","time":1507736103222,"user_id":"63920de858314afab527dfe3ea0ce86e","device_id":"abda2ca25c3843f6bf84fdbbf712a72b","event_type":"fxa_email - click","session_id":1507736080954,"event_properties":{"device_id":"abda2ca25c3843f6bf84fdbbf712a72b","email_type":"registration","service":"sync"},"user_properties":{"flow_id":"1a0b1cc2636b4f8198b05cbfb126a1aadda754c8c1e132490524f94897d3c7d8","experiments":["disabled_button_state_treatment","q3form_changes_disabled_button_state","send_sms_signin_codes"]},"app_version":"97","language":"en"}
{"op":"amplitudeEvent","time":1507736103317,"user_id":"63920de858314afab527dfe3ea0ce86e","device_id":"abda2ca25c3843f6bf84fdbbf712a72b","event_type":"fxa_sms - view","session_id":1507736080954,"event_properties":{"device_id":"abda2ca25c3843f6bf84fdbbf712a72b"},"user_properties":{"flow_id":"1a0b1cc2636b4f8198b05cbfb126a1aadda754c8c1e132490524f94897d3c7d8","experiments":["disabled_button_state_treatment","q3form_changes_disabled_button_state","send_sms_signin_codes"]},"app_version":"97","language":"en"}
```

@mozilla/fxa-devs, this is now ready for review. r?

File: server/lib/amplitude.js [hunk 1]
```diff
@@ -122,9 +122,9 @@ const EVENT_PROPERTIES = {
 
 const USER_PROPERTIES = {
   [GROUPS.email]: NOP,
-  [GROUPS.login]: mixProperties(mapUid, mapUtmProperties),
-  [GROUPS.registration]: mixProperties(mapUid, mapUtmProperties),
-  [GROUPS.settings]: mixProperties(mapUid, mapNewsletterState),
```
Comment (philbooth) [inline server/lib/amplitude.js]: This `mapUid` function was a relic so I got rid.

**Comment by davismtl** [comment]
@philbooth is the user property set at the same time as an event?

**Comment by philbooth** [comment]
> is the user property set at the same time as an event?

IIUC, yes. It's sent on events from the content server to the HTTP API, not separately out via the Identify API. Is that what you meant @davismtl?

**Comment by davismtl** [comment]
Yes, that's exactly what I meant. Awesome! :D

**Pull request merged**